### PR TITLE
Pawn promotion

### PR DESCRIPTION
**Pull request by yshrdbrn**
I handled the pawn promotion when we get it as a PGN.
For handling it on our side though, it needs the AI to be there.
I had class on Monday so I couldn't come to the coding night session so I don't know if you guys talked about it or not, but we should discuss the structure of it.

### CODE REVIEW

File: converters/PGNConverter.cpp [hunk 1]
```diff
@@ -105,9 +113,41 @@ int PGNConverter::getAlgebraicMoveType(std::string algebraicString) {
         return PAWN_MOVE;
     }
 
+    std::regex pawnPromotionRegEx("=[NBQR]");
+    if (std::regex_search(algebraicString, sm, pawnPromotionRegEx))
```
Comment (benrs) [inline converters/PGNConverter.cpp]: This will never be reached, PAWN_MOVE already checks for `(=[NBQR])?` in it's regex

Comment (benrs) [reply converters/PGNConverter.cpp]: You can do another regex under PAWN_MOVE to check if it's promoted and if it is then you can call you promotion function after the move.

**Comment by yshrdbrn** [comment]
I fixed the error Ben mentioned by moving the pawnPromotion code above the pawnMove

**Comment by jacobrs** [comment]
This PR does not include the Pawn promotion as part of the possibleMoves function of the Pawn class right? That is something we might want to add as part of this PR. This would also allow you to add some unit tests for your code.

**Comment by yshrdbrn** [comment]
Yeah, the problem is that the structure of possibleMoves can't handle the promotion right now since it only has the final destination for each piece. We need to define how the AI works in order to know how to define the possibleMoves.

File: converters/PGNConverter.cpp [hunk 2]
```diff
@@ -64,12 +64,20 @@ void PGNConverter::playAlgebraicMove(Color currPlayer, std::string algebraicStri
...
+            pawnPromotionMove(startingPosition, endingPosition, sm.str(0), board);
+            break;
+        }
         default:
             // Throw an exception since the pgn format is invalid
             throw std::invalid_argument("The pgn passed in is invalid.");
     }
 
-    if (!board->movePiece(startingPosition, endingPosition, currPlayer)) {
+    if (moveType != PAWN_PROMOTION && !board->movePiece(startingPosition, endingPosition, currPlayer)) {
```
Comment (benrs) [inline converters/PGNConverter.cpp:80]: Not sure if this really feels intuitive. I feel like movePiece should potentially be able to handle the promotion of a pawn, that way we don't need this special case for 'PAWN_PROMOTION' and we simply move the piece like normal + pass in an extra parameter letting the move know what promotion value we want.

Comment (benrs) [reply converters/PGNConverter.cpp:80]: Let me know if you have any criticisms about this method of tackling the problem.

Comment (benrs) [reply converters/PGNConverter.cpp:80]: Or even create a proxy method movePieceAndPromote.

**Comment by jacobrs** [comment]
We will need to refactor the PossibleMoves function to include those special moves such as promotions, castling and en-passant. Also tests should be writeable if this issue is to be merged since there is no way to use the code currently.

**Review by jacobrs — CHANGES_REQUESTED** [review]
We need to discuss refactoring possibleMoves.

**Pull request closed without merging**